### PR TITLE
Incremental Training for imagenet

### PR DESCRIPTION
**Pull request by qiuxin2012**
With incremental training, we can use less DRAM during the training.
The training workload is like this:
1. Persist origin data on disk
2. Sample a piece of data to memory
3. train these cached samples
4. unpersist these cached samples
5. Goto 2.

### CODE REVIEW

File: zoo/src/main/scala/com/intel/analytics/zoo/feature/pmem/NativeArray.scala [hunk 1]
```diff
@@ -32,6 +32,8 @@ case object PARTITIONED extends DataStrategy
 
 case object REPLICATED extends DataStrategy
 
+case class INCREMENTAL(cachePercentage: Double) extends DataStrategy
```
Comment (jason-dai) [inline zoo/src/main/scala/com/intel/analytics/zoo/feature/pmem/NativeArray.scala]: It makes more sense to make it a new `MemoryType` (e.g., DISK) instead of `DataStrategy`; e.g., we may later support DISK + REPLICATED

File: zoo/src/main/scala/com/intel/analytics/zoo/examples/inception/ImageNet2012.scala [hunk 2]
```diff
@@ -78,27 +77,24 @@ object ImageNet2012 {
...
-          batchSize = batchSize,
-          transformer = (BytesToBGRImg()
+      val featureSet = FeatureSet.rdd(rawData, memoryType = memoryType, dataStrategy)
+      featureSet.transform(BytesToBGRImg()
             -> BGRImgCropper(imageSize, imageSize)
             -> DatasetHFlip(0.5)
-            -> BGRImgNormalizer(0.485, 0.456, 0.406, 0.229, 0.224, 0.225))
-        ))
+            -> BGRImgNormalizer(0.485, 0.456, 0.406, 0.229, 0.224, 0.225)
+            -> BGRImgToSample()
```
Comment (jason-dai) [inline zoo/src/main/scala/com/intel/analytics/zoo/examples/inception/ImageNet2012.scala]: Why add toSample and toBatch here?

File: zoo/src/main/scala/com/intel/analytics/zoo/feature/FeatureSet.scala [hunk 3]
```diff
@@ -295,6 +296,58 @@ class CachedDistributedFeatureSet[T: ClassTag]
   }
 }
 
+/**
+ * Wrap a RDD as a FeatureSet.
+ * @param buffer
+ */
+// T is the returning value type. like ByteRecord
+class IncrementalFeatureSet[T: ClassTag]
```
Comment (jason-dai) [inline zoo/src/main/scala/com/intel/analytics/zoo/feature/FeatureSet.scala]: DiskFeatureSet?

File: zoo/src/main/scala/com/intel/analytics/zoo/feature/FeatureSet.scala [hunk 4]
```diff
@@ -295,6 +296,58 @@ class CachedDistributedFeatureSet[T: ClassTag]
...
+    if (train) {
+      if (currentFeatureSet != null) {
+        currentFeatureSet.unpersist()
+      }
+      currentSlice = buffer.sample(false, cachePercentage)
+        .setName(s"${cachePercentage * 100}% of ${origin.name}")
+      currentFeatureSet = DRAMFeatureSet.rdd(currentSlice)
+      currentFeatureSet.cache()
+      currentFeatureSet.data(train)
+    } else {
```
Comment (jason-dai) [inline zoo/src/main/scala/com/intel/analytics/zoo/feature/FeatureSet.scala]: What does it mean? In prediction/evaluation, only data in `currentFeatureSet` are used?

File: zoo/src/main/scala/com/intel/analytics/zoo/feature/FeatureSet.scala [hunk 5]
```diff
@@ -295,6 +296,58 @@ class CachedDistributedFeatureSet[T: ClassTag]
...
+    .setName("Origin Data Cached on Disk")
+  protected lazy val count: Long = buffer.count()
+
+  protected var currentSlice = buffer.sample(false, cachePercentage)
+  protected var currentFeatureSet: DistributedFeatureSet[T] = FeatureSet.rdd(currentSlice)
+
+  override def data(train: Boolean): RDD[T] = {
+    if (train) {
+      if (currentFeatureSet != null) {
+        currentFeatureSet.unpersist()
```
Comment (jason-dai) [inline zoo/src/main/scala/com/intel/analytics/zoo/feature/FeatureSet.scala]: Why do you init `currentFeatureSet` above and then discard it immediately here? maybe just init `currentFeatureSet` to null here?

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/models/Topology.scala [hunk 6]
```diff
@@ -1005,7 +1239,38 @@ private[zoo] class InternalDistriOptimizer[T: ClassTag] (
...
+        state("epoch") = trueEpoch
+        val newEndWhen = Trigger.or(endWhen, Trigger.maxIteration(i * iterations),
+          Trigger.maxEpoch(trueEpoch))
+        this.setEndWhen(newEndWhen)
+        if (checkPointTrigger.isDefined && checkpointDir.isDefined) {
+          // we should setCheckpoint every time before we call optimize(),
+          // as BigDL will overwrite checkpointPath to its subfolder.
+          this.setCheckpoint(checkpointDir.get, checkPointTrigger.get)
+        }
+        this.train()
```
Comment (jason-dai) [inline zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/models/Topology.scala]: We should set `state("epoch")` to `trueEpoch` immediately after train, otherwise `endWhen(state)` can exit the loop too early?

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/models/Topology.scala [hunk 7]
```diff
@@ -955,8 +999,197 @@ private[zoo] class InternalDistriOptimizer[T: ClassTag] (
     _criterion: Criterion[T])
   (implicit ev: TensorNumeric[T]) extends DistriOptimizer[T](_model, _dataset, _criterion)
   with AbstractEstimator[T]{
-  import InternalDistriOptimizer.logger
+  import InternalDistriOptimizer._
   protected var checkpointDir: Option[String] = None
+  protected var cachePercentage: Double = 1.0
```
Comment (jason-dai) [inline zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/models/Topology.scala]: I don't think you should make `cachePercentage` a variable in `Estimator`; instead, make it part of `FeatureSet`, and `Estimator` can then check its value in train.

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/models/Topology.scala [hunk 8]
```diff
@@ -955,8 +999,197 @@ private[zoo] class InternalDistriOptimizer[T: ClassTag] (
...
+            throw t
+          }
+      }
+    }
+
+    InternalOptimizerUtil.getModel(
+      cachedModels, parameters, trainingModel)
+
+    trainingModel
+  }
```
Comment (jason-dai) [inline zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/models/Topology.scala]: Do we need to call `optimizer.close()` here?

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/models/Topology.scala [hunk 9]
```diff
@@ -955,8 +999,197 @@ private[zoo] class InternalDistriOptimizer[T: ClassTag] (
...
+
+  override def close(): Unit = {
```
Comment (jason-dai) [inline zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/models/Topology.scala]: Shall we call `close` automatically when `train` is done?

File: zoo/src/main/scala/com/intel/analytics/zoo/examples/inception/ImageNet2012.scala (hunk 2 above)
Comment (qiuxin2012) [reply zoo/src/main/scala/com/intel/analytics/zoo/examples/inception/ImageNet2012.scala]: As the origin rdd is persisted on disk, when count the size of the dataset, MTLabeledBGRImgToBatch will through an exception.

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/models/Topology.scala (hunk 8 above)
Comment (qiuxin2012) [reply zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/models/Topology.scala]: optimizer.shutdown integrate to close()

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/estimator/Estimator.scala [hunk 10]
```diff
@@ -134,15 +121,22 @@ class Estimator[T: ClassTag] private[zoo](
...
-        case d: DistributedFeatureSet[MiniBatch[T]] =>
-          new InternalDistriOptimizer[T](model, null, criterion)
+    trainSet match {
+      case d: DistributedFeatureSet[MiniBatch[T]] =>
+        if (internalEstimator == null) {
+          internalEstimator = new InternalDistriOptimizer[T](model, null, criterion)
             .setCheckpointDir(modelDir)
             .setOptimMethods(optimMethods)
+        }
+        if (d.originSet().isInstanceOf[DiskFeatureSet[Any]]) {
```
Comment (jason-dai) [inline zoo/src/main/scala/com/intel/analytics/zoo/pipeline/estimator/Estimator.scala]: maybe we can define a method `memoryPercentage` in `FeatureSet` which returns 1 by default, and returns `cachePercentage` in DiskFeatureSet, and then you can just call this method inside `internalEstimator.train`

**Comment by qiuxin2012** [comment]
Add ZooTrigger to fix every epoch won't stop at correct iteration.

File: zoo/src/main/scala/com/intel/analytics/zoo/feature/FeatureSet.scala [hunk 11]
```diff
@@ -295,6 +304,90 @@ class CachedDistributedFeatureSet[T: ClassTag]
...
+          "because numSlice is zero. numSlice should >= 2 " +
+          "in a training FeatureSet.")
+      } else {
+        buffer
+      }
+    } else {
+      if (train) {
+        if (currentFeatureSet != null && trained) {
+          currentFeatureSet.unpersist()
+          newSample()
```
Comment (zhichao-li) [inline zoo/src/main/scala/com/intel/analytics/zoo/feature/FeatureSet.scala]: Wondering does the user need to enlarge the training epoch accordingly as it seems like the epoch is calculated as per num slice?

Comment (qiuxin2012) [reply zoo/src/main/scala/com/intel/analytics/zoo/feature/FeatureSet.scala]: We will reset the epoch number if the epoch is not ended actually.

File: zoo/src/main/scala/com/intel/analytics/zoo/feature/FeatureSet.scala [hunk 12]
```diff
@@ -295,6 +304,90 @@ class CachedDistributedFeatureSet[T: ClassTag]
...
+        }
+        currentFeatureSet.cache()
+        trained = true
```
Comment (cyita) [inline zoo/src/main/scala/com/intel/analytics/zoo/feature/FeatureSet.scala]: CurrentFeatureSet should be shuffled here.

File: zoo/src/main/scala/com/intel/analytics/zoo/common/ZooTrigger.scala [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+    this.zooState = zooState
+  }
+}
+
+/**
+ * A trigger that triggers an action when each epoch finishs.
+ * Could be used as trigger in setValidation and setCheckpoint
+ * in Optimizer, and also in TrainSummary.setSummaryTrigger.
+ */
+case class EveryEpoch() extends ZooTrigger{
```
Comment (cyita) [inline zoo/src/main/scala/com/intel/analytics/zoo/common/ZooTrigger.scala]: If memoryType DRAM is used, the validation phase is ignored at the end of each epoch.

Comment (qiuxin2012) [reply zoo/src/main/scala/com/intel/analytics/zoo/common/ZooTrigger.scala]: fixed

File: zoo/src/main/scala/com/intel/analytics/zoo/examples/inception/ImageNet2012.scala (hunk 2 above)
Comment (qiuxin2012) [reply zoo/src/main/scala/com/intel/analytics/zoo/examples/inception/ImageNet2012.scala]: reverted, As I won't count the hold rdd on disk.

**Comment by qiuxin2012** [comment]
jenkins passed
